### PR TITLE
Compress atomic class names

### PR DESCRIPTION
**Pull request by liamqma**
This PR does

- updates ax function to accept short class name
- updates babel-plugin to compress class names based on pre-bundling hash map
- generates compression map

### CODE REVIEW

**Comment by liamqma** [comment]
TODO

- [ ] Changeset
- [ ] Parcel integration

**Review by pancaspe87 — COMMENTED** [review]
Are we testing overrides here?

File: packages/react/src/runtime/__tests__/ax.test.ts
Comment (pancaspe87) [inline packages/react/src/runtime/__tests__/ax.test.ts:13]: Are we testing overrides here?

File: packages/react/src/runtime/__tests__/ax.test.ts
Comment (pancaspe87) [inline packages/react/src/runtime/__tests__/ax.test.ts:18]: I haven't followed up the previous PRs and I am still wrapping my head around these changes. Hence, apologies if this was mentioned somewhere else

I am assuming that:
- `_aaaabbbb` is the CSS class with no manipulation
- `a` is the optimised class, which is stored in the map
- `_aaaa_a` is made of atomic group + optimised class -> in this way we are able to handle overrides

File: examples/webpack/class-name-compression-map.json
Comment (pancaspe87) [inline examples/webpack/class-name-compression-map.json:2]: Where/how does this map get generated?

File: packages/babel-plugin/src/__tests__/index.test.ts
Comment (pancaspe87) [inline packages/babel-plugin/src/__tests__/index.test.ts:204]: Do we still have cases where compression is not safe, hence we should bail out?

File: packages/react/src/runtime/__tests__/ax.test.ts
Comment (liamqma) [reply packages/react/src/runtime/__tests__/ax.test.ts:18]: Yes, spot on!

File: packages/react/src/runtime/__tests__/ax.test.ts
Comment (liamqma) [reply packages/react/src/runtime/__tests__/ax.test.ts:13]: I think so. It should be covered by the next test case.
```
[
      'should ensure the last atomic declaration of a single group with short class name wins',
      ['_aaaabbbb', '_aaaacccc', '_aaaa_a'],
      'a',
],
```
Is it what you refer to?

File: examples/webpack/class-name-compression-map.json
Comment (liamqma) [reply examples/webpack/class-name-compression-map.json:2]: Good question! For now it's assumed that library consumers manually curate the map and pass it to the babel-plugin. We will later create an automatic method to generate the map by statically analysing the codebase, and the map should be periodically updated by some a build.

File: packages/babel-plugin/src/class-names/index.ts
Comment (liamqma) [inline packages/babel-plugin/src/class-names/index.ts:135]: The change in this file is to make <ClassNames> handle conditional CSS rules.
i.e. 
```
<ClassNames>
          {({ css }) => (<div className={css({ ...props.isPrimary && { color: 'blue' }})}>hello, world!</div>)}
</ClassNames>
```

File: packages/css/src/plugins/atomicify-rules.ts
Comment (liamqma) [inline packages/css/src/plugins/atomicify-rules.ts:93]: If compressed class name is available, we are replacing full class name with the compressed in the sheet.
So
```
.aaaabbbb { font-size: 10px }
```
becomes
```
.a { font-size: 10px }
```
It will reduce the size of stylesheet but will break if we don't replace every occurrence of `_aaaabbbb`, which I should have done in this PR. 🤞

File: packages/babel-plugin/src/__tests__/index.test.ts
Comment (liamqma) [reply packages/babel-plugin/src/__tests__/index.test.ts:204]: Compression should be safe everywhere now.

File: packages/react/src/runtime/__tests__/ax.test.ts
Comment (pancaspe87) [reply packages/react/src/runtime/__tests__/ax.test.ts:13]: yep!

File: packages/css/src/utils/class-name-generator.ts
Comment (JakeLane) [inline packages/css/src/utils/class-name-generator.ts:1]: Did you miss uppercase chars?

File: packages/babel-plugin/src/__tests__/index.test.ts
Comment (pancaspe87) [inline packages/babel-plugin/src/__tests__/index.test.ts:221]: Does the extracted CSS also contain
`._1wyb_a, .1wyb1fwx {font-size:12px}`?

File: packages/css/src/__tests__/generate-compression-map.test.ts
Comment (pancaspe87) [inline packages/css/src/__tests__/generate-compression-map.test.ts:123]: Is the compression map supposed to be stable across builds? 
Here we are just testing that it gets generated; however, in the following test it seems that a different compressed class is generated for `17gjpfqs`

File: packages/css/src/utils/class-name-generator.ts
Comment (liamqma) [reply packages/css/src/utils/class-name-generator.ts:1]: Good point. I had a brief look at the spec. CSS class is case sensitive. We'll do more research to confirm it.

File: packages/babel-plugin/src/__tests__/index.test.ts
Comment (liamqma) [reply packages/babel-plugin/src/__tests__/index.test.ts:221]: No, the extracted CSS looks like
```
.a {font-size:12px}
```

File: packages/react/src/runtime/__perf__/ax.test.ts
Comment (dddlr) [inline packages/react/src/runtime/__perf__/ax.test.ts:43]: Nit: typo in "compresssed"

File: packages/css/src/__tests__/generate-compression-map.test.ts
Comment (liamqma) [reply packages/css/src/__tests__/generate-compression-map.test.ts:123]: I hope I understand the question correctly 🙂. Let me explain it.

> Is the compression map supposed to be stable across builds?

Yes, the compression map is supposed to be stable across builds. The way it achieves this is to take the compression map from the previous build as parameter:

```
generateCompressionMap = (
  css: string,
  oldClassNameCompressionMap?: { [index: string]: string }
):
```

For example, let's say we have the CSS:
```
._aaaabbbb { color: red }
._aaaacccc { color: black }
```
oldClassNameCompressionMap
```
{
   "aaaabbbb": "ddd"
}
```
`generateCompressionMap` returns new compression map
```
{
   "aaaacccc": "a"
   "aaaabbbb": "ddd" // Force "aaaabbbb" to be "ddd", which is same as oldClassNameCompressionMap
}
```

In the following test case, without providing an oldCompressionMap, '17gjpfqs' is compressed to "yb", but because we provide oldCompressionMap as  { "17gjpfqs": "a", 1vdp1hna': 'b',  '14fy1hna': 'c'}, '17gjpfqs' is compressed to 'a'. We want to ensure the new compression map keeps the same class name mapping from the oldCompressionMap.

File: examples/parcel/.compiledcssrc
Comment (JakeLane) [inline examples/parcel/.compiledcssrc:24]: this should probably be an external file right?

File: packages/babel-plugin/src/__tests__/index.test.ts
Comment (JakeLane) [inline packages/babel-plugin/src/__tests__/index.test.ts:289]: nit: inconsistent quotes?

File: packages/css/src/__tests__/generate-compression-map.test.ts
Comment (pancaspe87) [reply packages/css/src/__tests__/generate-compression-map.test.ts:123]: thanks for the explanation! Now I get it! I was looking at the  key/pair values that I didn't realise what the comparison was about  :)

File: packages/css/src/generate-compression-map.ts
Comment (pancaspe87) [inline packages/css/src/generate-compression-map.ts:22]: Nice! 

Out of curiosity, was this based on Tom's comment or did you find any resource with a list of AdBlocker classes.
Maybe `adv` could be an issue too ( [eg](https://stackoverflow.com/a/20242141))? 

Also, given that we prefix classes with `_`, I wonder if compressed classes are safe, regardless

File: packages/react/src/runtime/__perf__/ax.test.ts
Comment (pancaspe87) [reply packages/react/src/runtime/__perf__/ax.test.ts:43]: good catch!

File: packages/css/src/utils/__tests__/class-name-generator.test.ts
Comment (pancaspe87) [inline packages/css/src/utils/__tests__/class-name-generator.test.ts:18]: I guess this is safe because the class will be prefixed with `_`

File: packages/react/src/runtime/__perf__/ax.test.ts
Comment (pancaspe87) [inline packages/react/src/runtime/__perf__/ax.test.ts:20]: Why is there `undefined` in here?

File: packages/css/src/__tests__/generate-compression-map.test.ts
Comment (dddlr) [inline packages/css/src/__tests__/generate-compression-map.test.ts:48]: Would it make a difference if we change all of these keys (irr3k6hx, jomrk6hx, etc) to be in single quotes? I'm not quite sure why some are in single quotes and some are not

Comment (dddlr) [reply packages/css/src/__tests__/generate-compression-map.test.ts:48]: Ohh I think I see why, the number(s) at the start of the key means that the single quotes are required right?

Comment (dddlr) [reply packages/css/src/__tests__/generate-compression-map.test.ts:48]: Would personally prefer consistency, i.e. all of the keys being in single quotes

File: packages/css/src/generate-compression-map.ts
Comment (liamqma) [reply packages/css/src/generate-compression-map.ts:22]: Good point! I had a look at the easylist. I think I will take the approach of [styled-components](https://github.com/styled-components/styled-components/blob/main/packages/styled-components/src/utils/generateAlphabeticName.ts#L1), which avoids any word includes `ad`. Let me update the code.

Comment (dddlr) [reply packages/css/src/generate-compression-map.ts:22]: Should we add 'ddb' to this list as well? [EasyList](https://easylist.to/easylist/easylist.txt) (which is available for Adblock Plus, uBlock Origin etc) will block any class named "ddb"

> Also, given that we prefix classes with _, I wonder if compressed classes are safe, regardless

I'm curious about this too :)

File: packages/css/src/utils/__tests__/class-name-generator.test.ts
Comment (liamqma) [reply packages/css/src/utils/__tests__/class-name-generator.test.ts:18]: Ah, good catch. The test case should be `should NOT generate class names starting with a number`. 🤦‍♂️
I still haven't made the mind of whether prefix with `_`. Prefixing with `_` guarantee the uniqueness but reduce the perf improvement. 
I think I will provide a option and let the library consumer to decide.

File: packages/react/src/runtime/__perf__/ax.test.ts
Comment (liamqma) [reply packages/react/src/runtime/__perf__/ax.test.ts:20]: `undefined` is an acceptable parameter so we want to include it in the test case.

```
ax([undefined, '.aaaabbbb'])
```

btw, I didn't add `undefined`.

File: packages/css/src/__tests__/generate-compression-map.test.ts
Comment (liamqma) [reply packages/css/src/__tests__/generate-compression-map.test.ts:48]: The formatting is automatically done by [prettier](https://prettier.io/blog/2019/04/12/1.17.0.html#add-an-option-to-modify-when-prettier-quotes-object-properties-5934-by-azz). We don't customize this rule so it's the default "as-needed". I am not particular about the formatting but happy to change it if you prefer.

File: packages/babel-plugin/src/__tests__/index.test.ts
Comment (liamqma) [reply packages/babel-plugin/src/__tests__/index.test.ts:289]: https://github.com/atlassian-labs/compiled/pull/1408#discussion_r1106683169

File: examples/parcel/.compiledcssrc
Comment (liamqma) [reply examples/parcel/.compiledcssrc:24]: - [ ] externalize the map

File: packages/css/src/__tests__/generate-compression-map.test.ts
Comment (dddlr) [reply packages/css/src/__tests__/generate-compression-map.test.ts:48]: Ah no worries, thank you for clarifying! Since it's being done by our linter, I'd be happy to leave it as-is

File: packages/css/src/utils/__tests__/class-name-generator.test.ts
Comment (pancaspe87) [reply packages/css/src/utils/__tests__/class-name-generator.test.ts]: +1 on making it optional, sounds a perfect solution for different consumers :)

File: packages/react/src/runtime/__perf__/ax.test.ts
Comment (dddlr) [reply packages/react/src/runtime/__perf__/ax.test.ts:20]: Optional: I know the `undefined` wasn't added by your PR, but I think incorporating what you've said here (`undefined` is a valid parameter to `ax`) into a brief comment in the code itself might be helpful for future developers. I know that I'd definitely be confused to see it in a test case as well 😅

Comment (liamqma) [reply packages/react/src/runtime/__perf__/ax.test.ts:20]: - [ ] Add comment

**Comment by dddlr** [comment]
Thank you Liam for updating the PR :) This looks good to me, but I'll wait for an approval from @JakeLane or @pancaspe87 too since I don't have a good understanding of the whole codebase

**Review by JakeLane — APPROVED** [review]

**Review by dddlr — APPROVED** [review]

**Pull request merged**
